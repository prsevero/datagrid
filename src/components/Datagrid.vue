<template>
    <table>
        <thead v-if="headers && itemKeys.length">
            <tr>
                <th v-for="(key, index) in itemKeys" :key="index" :style="key.style" @click="sort(key)">
                    <span v-if="!keyIsObject(key)">{{ headers[key] }}</span>
                    <span v-else>{{ headers[key.key] }}</span>
                    <img src="./arrow.svg" v-if="isSorted(key)" :class="{desc: sortedBy.desc}" />
                </th>
            </tr>

            <tr>
                <th v-for="(key, index) in itemKeys" :key="index">
                    <span v-if="!keyIsObject(key)">
                        <input placeholder="Filter" type="text" v-model="filters[key]" @input="filter(key)" />
                    </span>
                    <span v-else>
                        <input placeholder="Filter" type="text" v-model="filters[key.key]" @input="filter(key.key)" />
                    </span>
                </th>
            </tr>
        </thead>

        <tbody>
            <tr v-for="(item, index) in internalItems" :key="index">
                <td v-for="(key, index) in itemKeys" :key="index" :style="[key.style, key.tdStyle]">
                    <span v-if="!keyIsObject(key)">{{ item[key] }}</span>
                    <span v-else>
                        <input type="text"
                            v-if="key.editable"
                            v-model="item[key.key]"
                            @blur="onSave"
                            @input="onInput"
                            @keyup.enter="onSave" />
                        <span v-else>{{ call(key, item) }}</span>
                    </span>
                </td>
            </tr>
        </tbody>
    </table>
</template>


<script>
import moment from 'moment';

let isFiltered = false,
    shouldSave = false;

export default {
    name: 'Datagrid',
    data() {
        return {
            internalItems: [],
            filters: [],
            sortedBy: {},
        };
    },
    props: {
        headers: Array,
        items: {
            required: true,
            type: Array,
        },
        itemKeys: {
            required: true,
            type: Array,
        },
        save: Function,
    },
    computed: {
        call() {
            return (key, item) => {
                if (key.fn === 'format') {
                    if (key.type === 'date')
                        return this.formatDate(item[key.key], key.format);
                }

                return item[key.key];
            };
        },
        formatDate() {
            return (date, format = 'DD/MM/YYYY HH:mm:ss') =>
                moment(date).format(format);
        },
        keyIsObject() {
            return key => typeof key === 'object';
        },
    },
    methods: {
        isSorted(key) {
            return (
                key === this.sortedBy.key ||
                (this.sortedBy.key && key.key === this.sortedBy.key)
            );
        },
        onInput() {
            shouldSave = true;
        },
        onSave() {
            if (shouldSave) {
                shouldSave = false;
                this.save(this.internalItems);
            }
        },
        filter() {
            let itemsToFilter = this.items;
            isFiltered = false;

            this.filters.map((filter, i) => {
                if (filter !== undefined) {
                    isFiltered = true;
                    filter =
                        typeof filter === 'string'
                            ? filter.toLowerCase()
                            : filter;

                    this.internalItems = itemsToFilter.filter(item => {
                        let value = item[i];
                        if (typeof value === 'string') {
                            value = value.toLowerCase();
                            return value.indexOf(filter) >= 0;
                        } else return value === filter;
                    });
                    itemsToFilter = this.internalItems;
                }
            });

            if (!isFiltered) this.internalItems = [...this.items];
            this.sort(this.sortedBy, false);
        },
        sort(by, changeOrder = true) {
            if (by !== undefined) {
                if (!this.keyIsObject(by)) by = {key: by};

                const isNumber =
                    by.type &&
                    (by.type === 'float' ||
                        by.type === 'int' ||
                        by.type === 'number');

                this.internalItems.sort((valueA, valueB) => {
                    let a = valueA[by.key],
                        b = valueB[by.key];

                    if (isNumber) {
                        a = parseFloat(a);
                        b = parseFloat(b);
                    }

                    return a > b ? 1 : a < b ? -1 : 0;
                });

                if (changeOrder)
                    by.desc =
                        by.key === this.sortedBy.key && !this.sortedBy.desc;
                if (by.desc) this.internalItems.reverse();

                this.sortedBy = by;
                if (!isFiltered) this.save(this.internalItems);
            }
        },
    },
    mounted() {
        this.internalItems = [...this.items];
    },
};
</script>


<style scoped lang="sass">
table
    border: 0
    border-collapse: collapse
    font-size: 14px
    position: relative
    width: 100%

    // Sticky headers
    th
        position: sticky
    thead tr
        &:first-child th
            top: 50px
        &:last-child th
            padding-top: 0
            top: 82px
    // /Sticky headers

    td
        padding: 8px

    thead, th
        background: #000
    tbody
        background: #ddd
    th
        color: #fff
        padding: 8px 8px 12px
        text-align: left
    thead tr:first-child th
        cursor: pointer
        padding-top: 15px
        img
            display: inline-block
            margin-left: 4px
            vertical-align: middle
            width: 12px
            transform: rotate(90deg)
            &.desc
                transform: rotate(-90deg)

    tbody tr
        &:nth-child(odd),
        &:nth-child(odd) input
            background: #f7f7f7
        &:nth-child(odd) input
            border-color: #f7f7f7
        &:nth-child(even),
        &:nth-child(even) input
            background: #eaeaea
        &:nth-child(even) input
            border-color: #eaeaea
        &:hover,
        &:hover input
            background: #ddd
        &:hover input
            border-color: #ddd

    input
        appearance: none
        border: 1px solid transparent
        border-radius: 0
        outline: 0
        padding: 4px 6px
        text-overflow: ellipsis
        width: 100%

    tbody tr:nth-child(odd) td input,
    tbody tr:nth-child(even) td input
        cursor: pointer
        @media (pointer: coarse)
            border-color: #ccc
        &:focus
            background: #fff
            border-color: #bbb
            transition: all .22s ease-in-out
</style>
