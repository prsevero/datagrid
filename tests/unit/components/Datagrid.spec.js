import { shallowMount } from '@vue/test-utils';
import Datagrid from '@/components/Datagrid.vue';
import moment from 'moment';


describe('Datagrid.vue', () => {

    let props;
    let mountedDatagrid;
    const datagrid = () => {
        if (!mountedDatagrid) {
            mountedDatagrid = shallowMount(Datagrid, {propsData: props});
        }
        return mountedDatagrid;
    }

    beforeEach(() => {
        props = {
            itemKeys: [],
            items: [],
            save: jest.fn(),
        };
        mountedDatagrid = undefined;
    });


    // All tests will go here
    it('should match snapshot', () => {
        const wrapper = datagrid();
        expect(wrapper.vm.$el).toMatchSnapshot();
    });


    it('always render a datagrid', () => {
        const wrapper = datagrid();
        expect(wrapper.is('table')).toBe(true);
    });


    describe('Headers', () => {
        it('should not render headers if it is not passed', () => {
            const wrapper = datagrid();
            expect(wrapper.findAll('thead').length).toBe(0);
        });


        it('should not render headers if it is passed, but do not has itemKeys', () => {
            const headers = [0, 'ID', {key: 0}, {key: 'ID'}];
            props.headers = headers;
            const wrapper = datagrid();
            expect(wrapper.findAll('thead').length).toBe(0);
        });


        it('should render headers and filters correctly if it is passed the headers and itemKeys', () => {
            const headers = [0, 'ID', 2, '3'];
            const itemKeys = [0, 1, {key: 2}, {key: 3}];
            props.headers = headers;
            props.itemKeys = itemKeys;

            const wrapper = datagrid();
            expect(wrapper.findAll('thead').length).toBe(1);
            expect(wrapper.findAll('tr:first-child th').length).toBe(headers.length);
            expect(wrapper.findAll('tr:last-child th').length).toBe(headers.length);
            expect(wrapper.findAll('tr:last-child input').length).toBe(headers.length);
        });


        it('should render headers with the passed values if it is passed the headers and itemKeys', () => {
            const headers = [0, 'ID', '2'];
            const itemKeys = [0, 1, {key: 2}, 3];
            props.headers = headers;
            props.itemKeys = itemKeys;

            const wrapper = datagrid();
            const ths = wrapper.findAll('tr:first-child th');
            expect(ths.at(0).text()).toBe(`${headers[0]}`);
            expect(ths.at(1).text()).toBe(headers[1]);
            expect(ths.at(2).text()).toBe(`${headers[2]}`);
        });


        it('should render the style if it is passed', () => {
            const headers = [0, 'ID', '2'];
            const itemKeys = [
                {key: 0, style: {textAlign: 'center'}},
                1,
                {key: 2, style: {textAlign: 'center'}},
            ];
            props.headers = headers;
            props.itemKeys = itemKeys;
            const wrapper = datagrid();

            const ths = wrapper.findAll('thead tr:first-child th');
            expect(ths.at(0).attributes().style).toBe('text-align: center;');
            expect(ths.at(1).attributes().style).toBeUndefined();
            expect(ths.at(2).attributes().style).toBe('text-align: center;');
        });
    });


    describe('Body', () => {

        it('should always render tbody', () => {
            const wrapper = datagrid();
            expect(wrapper.findAll('tbody').length).toBe(1);
        });


        it('should not render any row in tbody if there is no items', () => {
            const wrapper = datagrid();
            expect(wrapper.findAll('tbody tr').length).toBe(0);
        });


        it('tbody rows should match the items length when itemKeys is also passed', done => {
            const itemKeys = [0, 0, 0, 0];
            const items = [[0], [1], [2], ['3']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                expect(wrapper.findAll('tbody tr').length).toBe(items.length);
                done();
            }, 100);
        });


        it('should handle both index as integer or object', done => {
            const itemKeys = [0, {key: 0}];
            const items = [['ID'], ['Name']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).find('td').text()).toBe(items[0][itemKeys[0]]);
                expect(trs.at(1).find('td').text()).toBe(items[1][itemKeys[1].key]);
                done();
            }, 100);
        });


        it('should format a date correctly', done => {
            const today = moment();
            const tomorrow = today.clone().add(1, 'day');
            const format = 'YYYY-MM-DD';

            const itemKeys = [{fn: 'format', format: format, key: 0, type: 'date'}];
            const items = [[today.toDate()], [tomorrow.toDate()]];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).find('td').text()).toBe(today.format(format));
                expect(trs.at(1).find('td').text()).toBe(tomorrow.format(format));
                done();
            }, 100);
        });


        it('should parse a string as a number when sorting when the type is specified', done => {
            const itemKeys = [0];
            const items = [['1'], ['2.2']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                wrapper.vm.sort({key: 0, type: 'number'});
                const trs = wrapper.findAll('tbody tr');
                expect(parseFloat(trs.at(0).find('td').text())).toBe(1);
                expect(parseFloat(trs.at(1).find('td').text())).toBe(2.2);
                done();
            }, 100);
        });


        it('should render an input if the item index is editable', done => {
            const itemKeys = [0, {key: 1}, {key: 2, editable: true}];
            const items = [['ID', 'Name', 'Description'], ['ID', 'Name', 'Description']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('input').length).toBe(1);
                expect(trs.at(1).findAll('input').length).toBe(1);
                done();
            }, 100);
        });


        it('should call onSave function when blur or press enter in an editable input', done => {
            const itemKeys = [0, {key: 1}, {key: 2, editable: true}];
            const items = [['ID', 'Name', 'Description'], ['ID', 'Name', 'Description']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();
            wrapper.setMethods({onSave: jest.fn()});

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                const input = trs.at(0).find('input');
                expect(wrapper.vm.onSave).toHaveBeenCalledTimes(0);
                input.trigger('blur');
                expect(wrapper.vm.onSave).toHaveBeenCalledTimes(1);
                input.trigger('keyup.enter');
                expect(wrapper.vm.onSave).toHaveBeenCalledTimes(2);
                done();
            }, 100);
        });


        it('should render the td style if it is passed', done => {
            const itemKeys = [
                {key: 0, tdStyle: {textAlign: 'center'}},
                1,
                {key: 2, tdStyle: {textAlign: 'center'}},
            ];
            const items = [['ID', 'Name', 'Description']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).attributes().style).toBe('text-align: center;');
                expect(trs.at(0).findAll('td').at(2).attributes().style).toBe('text-align: center;');
                done();
            }, 100);
        });


        it('should render both style and td style if it is passed', done => {
            const itemKeys = [
                {key: 0, style: {fontSize: '20px'}, tdStyle: {textAlign: 'center'}},
                1,
                {key: 2, style: {fontSize: '20px'}, tdStyle: {textAlign: 'center'}},
            ];
            const items = [['ID', 'Name', 'Description']];
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).attributes().style).toBe('font-size: 20px; text-align: center;');
                expect(trs.at(0).findAll('td').at(2).attributes().style).toBe('font-size: 20px; text-align: center;');
                done();
            }, 100);
        });


        it('should sort things correctly by the parameter', done => {
            const headers = ['Name', 'Value'];
            const itemKeys = [0, 1];
            const items = [['Z', 10], ['P', -5], ['M', 0], ['A', 5]];
            props.headers = headers;
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const ths = wrapper.findAll('thead tr:first-child th');
                let trs = wrapper.findAll('tbody tr');

                // Original order
                expect(trs.at(0).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('10');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('P');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('-5');
                expect(trs.at(2).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(2).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(3).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(3).findAll('td').at(1).text()).toBe('5');

                // Order by name
                ths.at(0).trigger('click');
                trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('5');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(2).findAll('td').at(0).text()).toBe('P');
                expect(trs.at(2).findAll('td').at(1).text()).toBe('-5');
                expect(trs.at(3).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(3).findAll('td').at(1).text()).toBe('10');

                // Order by name descendent
                ths.at(0).trigger('click');
                trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('10');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('P');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('-5');
                expect(trs.at(2).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(2).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(3).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(3).findAll('td').at(1).text()).toBe('5');

                // Order by value
                ths.at(1).trigger('click');
                trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).text()).toBe('P');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('-5');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(2).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(2).findAll('td').at(1).text()).toBe('5');
                expect(trs.at(3).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(3).findAll('td').at(1).text()).toBe('10');

                // Order by value descendent
                ths.at(1).trigger('click');
                trs = wrapper.findAll('tbody tr');
                expect(trs.at(0).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('10');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('5');
                expect(trs.at(2).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(2).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(3).findAll('td').at(0).text()).toBe('P');
                expect(trs.at(3).findAll('td').at(1).text()).toBe('-5');

                done();
            }, 100);
        });


        it('should filter things correctly by the parameter', done => {
            const headers = ['Name', 'Value'];
            const itemKeys = [0, 1];
            const items = [['Z', 0], ['P', -5], ['M', 0], ['A', 5]];
            props.headers = headers;
            props.itemKeys = itemKeys;
            props.items = items;
            const wrapper = datagrid();

            setTimeout(() => {
                const ths = wrapper.findAll('thead tr:first-child th');
                let trs = wrapper.findAll('tbody tr');

                // Original data
                expect(trs.length).toBe(items.length);

                // Filter by name
                wrapper.setData({filters: ['a']});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(1);
                expect(trs.at(0).findAll('td').at(0).text()).toBe('A');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('5');

                // Filter by unexisting name
                wrapper.setData({filters: ['B']});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(0);

                // Filter by value
                wrapper.setData({filters: [, 0]});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(2);
                expect(trs.at(0).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('0');
                expect(trs.at(1).findAll('td').at(0).text()).toBe('M');
                expect(trs.at(1).findAll('td').at(1).text()).toBe('0');

                // Filter by unexisting value
                wrapper.setData({filters: [, 2]});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(0);

                // Filter by name and value
                wrapper.setData({filters: ['z', 0]});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(1);
                expect(trs.at(0).findAll('td').at(0).text()).toBe('Z');
                expect(trs.at(0).findAll('td').at(1).text()).toBe('0');

                // Filter by unexisting name and value
                wrapper.setData({filters: ['z', 1]});
                wrapper.find('thead input').trigger('input');
                trs = wrapper.findAll('tbody tr');
                expect(trs.length).toBe(0);

                done();
            }, 100);
        });

    });

});
