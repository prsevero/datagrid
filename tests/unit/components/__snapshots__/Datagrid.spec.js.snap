// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datagrid.vue should match snapshot 1`] = `
<table>
  <!---->
   
  <tbody />
</table>
`;
